TASVideos/TagHelpers/WikiMarkupTagHelper.cs

```diff
@@ -1,71 +1,126 @@
-			{
-				throw new InvalidOperationException($"Unknown ViewComponent: {name}");
-			}
-
-			var content = _viewComponentHelper.InvokeAsync(viewComponent, new { pageData = PageData, pp }).Result;
-			content.WriteTo(w, HtmlEncoder.Default);
-		}
-	}
-}
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Namotion.Reflection;
+using TASVideos.Data.Entity;
+using TASVideos.Extensions;
+using TASVideos.ViewComponents;
+using TASVideos.WikiEngine;
+using TASVideos.WikiEngine.AST;
+
+namespace TASVideos.TagHelpers
+{
+	public partial class WikiMarkup : TagHelper, IWriterHelper
+	{
+		private readonly IViewComponentHelper _viewComponentHelper;
+
+		public WikiMarkup(IViewComponentHelper viewComponentHelper)
+		{
+			_viewComponentHelper = viewComponentHelper;
+		}
+
+		[ViewContext]
+		[HtmlAttributeNotBound]
+		public ViewContext ViewContext { get; set; } = new ();
+
+		public string Markup { get; set; } = "";
+		public WikiPage PageData { get; set; } = new ();
+
+		public override void Process(TagHelperContext context, TagHelperOutput output)
+		{
+			((IViewContextAware)_viewComponentHelper).Contextualize(ViewContext);
+			output.TagName = "article";
+			output.AddCssClass("wiki");
+
+			var sw = new StringWriter();
+			Util.RenderHtmlDynamic(Markup, sw, this);
+			output.Content.SetHtmlContent(sw.ToString());
+		}
+
+		bool IWriterHelper.CheckCondition(string condition)
+		{
+			return HtmlExtensions.WikiCondition(ViewContext, condition);
+		}
+
+		private static readonly IDictionary<string, Type> ViewComponents = Assembly
+			.GetAssembly(typeof(WikiModuleAttribute))
+			!.GetTypes()
+			.Where(t => t.GetCustomAttribute(typeof(WikiModuleAttribute)) != null)
+			.ToDictionary(tkey => ((WikiModuleAttribute)tkey.GetCustomAttribute(typeof(WikiModuleAttribute))!).Name, tvalue => tvalue, StringComparer.InvariantCultureIgnoreCase);
+
+		void IWriterHelper.RunViewComponent(TextWriter w, string name, IReadOnlyDictionary<string, string> pp)
+		{
+			var componentExists = ViewComponents.TryGetValue(name, out Type? viewComponent);
+			if (!componentExists)
+			{
+				throw new InvalidOperationException($"Unknown ViewComponent: {name}");
+			}
+
+			var paramObject = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase)
+			{
+				{ "pageData", PageData }
+			};
+
+			var invokeMethod = viewComponent!.GetMethod("InvokeAsync")
+				?? viewComponent.GetMethod("Invoke");
+
+			if (invokeMethod == null)
+			{
+				throw new InvalidOperationException($"Could not find an Invoke method on ViewComponent {viewComponent}");
+			}
+
+			var paramCandidates = invokeMethod
+				.GetParameters()
+				.Where(p => !paramObject.ContainsKey(p.Name!)); // filter out any already supplied parameters
```

Comment:
Will this be case-insensitive when matching parameter names?  We need it to be, can't count on users to match variable camelCasing

Reply:
`paramObject` was initialized as `new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase)`, so this line is fine.

However, there's some question on the other side:  Will ASP.NET core's binder use the string equality I gave it?  I do not know, so leaving this open until I can check.

Reply:
I changed the frames viewcomponent's parameter list to `(WikiPage pagedata` instead of `(WikiPage pageData` and it still got the pagedata, and was able to figure out the canonical NES FPS on a NES game page.  So I think that's all we needed there, right?

Reply:
Oh, as far as the user supplied parameters go that all works too; `[module:frames|aMOUnt=400|    fPs=61.7]` is fully processed.